[BUG]: Backend test fails only locally - test_get_last_updated_by_human_ms()

### Describe the bug

This test of exp_services seems to be comparing the times of 2 actions that happen at almost the same time but one time is asserted to be less than another for some reason. Since the actions happen simultaneously, the test always fails locally although it doesn't fail on CI.


```
======================================================================
FAIL: test_get_last_updated_by_human_ms (core.domain.exp_services_test.ExplorationSnapshotUnitTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/apple/Desktop/oppia3/oppia/core/domain/exp_services_test.py", line 4842, in test_get_last_updated_by_human_ms
    self.assertLess(
AssertionError: 1723198956676.251 not less than 1723179156691.9639

----------------------------------------------------------------------
Ran 206 tests in 121.343s

FAILED (failures=1)

```

### Steps To Reproduce

1. Run the exp_services_test backend spec locally.

### Expected Behavior

The test should not fail locally, and supposedly the assertion itself is inappropriate. Needs investigation.

### What device are you using?

Desktop

### Operating System

MacOS

### Additional context

Link to the test function - https://github.com/oppia/oppia/blob/2932d9c65d157c7f0951bcfed714a67c87121a81/core/domain/exp_services_test.py#L4751

This PR might be helpful in finding a potential fix: #22582 

#### Current temporary fixes:

- Manually setting your system's timezone to UTC:

![Image](https://github.com/user-attachments/assets/067f3075-f66f-469a-99d6-24d3c427bd8e)

### Tips for developers

Before addressing the bug, please identify which PR caused the issue (you can follow the steps [here](https://github.com/oppia/oppia/wiki/How-to-find-the-commit-which-introduced-a-bug)). If you identify the PR, comment on the issue with a link to it. If not, mention the commit hash of the oldest commit you saw the bug on (and the month and year it was made in).

Then, please leave a comment with details of the approach that you plan to take to fix the issue (see [example](https://github.com/oppia/oppia/issues/19157#issuecomment-1858788463)).

**Note:** If this is your first Oppia issue, please make sure to follow our guidelines for [choosing an issue](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#choosing-a-good-first-issue) and [setting things up](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#setting-things-up). You will also need to show a demo of the fix working correctly on your local machine. Thanks!

--- a/core/domain/exp_services_test.py
+++ b/core/domain/exp_services_test.py
@@ -5859,23 +5859,13 @@ class ExplorationSnapshotUnitTests(ExplorationServicesUnitTests):
     SECOND_EMAIL: Final = 'abc123@gmail.com'
 
     def test_get_last_updated_by_human_ms(self) -> None:
-        original_timestamp = (
-            datetime.datetime.now(datetime.timezone.utc)
-            .replace(tzinfo=None)
-            .timestamp()
-            * 1000
-        )
+        original_timestamp = utils.get_current_time_in_millisecs()
 
         self.save_new_valid_exploration(
             self.EXP_0_ID, self.owner_id, end_state_name='End'
         )
 
-        timestamp_after_first_edit = (
-            datetime.datetime.now(datetime.timezone.utc)
-            .replace(tzinfo=None)
-            .timestamp()
-            * 1000
-        )
+        timestamp_after_first_edit = utils.get_current_time_in_millisecs()
 
         exp_services.update_exploration(
             feconf.MIGRATION_BOT_USER_ID,

--- a/core/utils.py
+++ b/core/utils.py
@@ -592,8 +592,13 @@ def get_time_in_millisecs(datetime_obj: datetime.datetime) -> float:
         float. The time in milliseconds since the Epoch.
     """
     if datetime_obj.tzinfo is None:
-        # Naive datetime from datastore. Treat it as UTC.
+        # Naive datetime (no tzinfo). Treated as UTC. This can come from
+        # NDB model fields (e.g. created_on, last_updated) which store
+        # naive UTC in the datastore.
         datetime_obj = datetime_obj.replace(tzinfo=datetime.timezone.utc)
+    else:
+        # Aware datetime. Normalize to UTC before converting.
+        datetime_obj = datetime_obj.astimezone(datetime.timezone.utc)
     return datetime_obj.timestamp() * 1000.0
 
 

--- a/core/utils.py
+++ b/core/utils.py
@@ -591,6 +591,9 @@ def get_time_in_millisecs(datetime_obj: datetime.datetime) -> float:
     Returns:
         float. The time in milliseconds since the Epoch.
     """
+    if datetime_obj.tzinfo is None:
+        # Naive datetime from datastore. Treat it as UTC.
+        datetime_obj = datetime_obj.replace(tzinfo=datetime.timezone.utc)
     return datetime_obj.timestamp() * 1000.0
 
 

--- a/core/utils_test.py
+++ b/core/utils_test.py
@@ -787,6 +787,32 @@ class UtilsTests(test_utils.GenericTestBase):
             ).replace(tzinfo=None),
         )
 
+    def test_get_time_in_millisecs_with_utc_aware_datetime(self) -> None:
+        dt = datetime.datetime(2020, 6, 15, tzinfo=datetime.timezone.utc)
+        msecs = utils.get_time_in_millisecs(dt)
+        self.assertEqual(
+            dt,
+            datetime.datetime.fromtimestamp(
+                msecs / 1000.0, datetime.timezone.utc
+            ),
+        )
+
+    def test_get_time_in_millisecs_with_non_utc_aware_datetime(self) -> None:
+        # A non-UTC aware datetime should be normalized to UTC correctly,
+        # not have tzinfo blindly stripped, which would corrupt the time.
+        ist = datetime.timezone(datetime.timedelta(hours=5, minutes=30))
+        dt_ist = datetime.datetime(2020, 6, 15, 5, 30, 0, tzinfo=ist)
+        dt_utc = datetime.datetime(
+            2020, 6, 15, 0, 0, 0, tzinfo=datetime.timezone.utc
+        )
+        msecs = utils.get_time_in_millisecs(dt_ist)
+        self.assertEqual(
+            dt_utc,
+            datetime.datetime.fromtimestamp(
+                msecs / 1000.0, datetime.timezone.utc
+            ),
+        )
+
     def test_convert_millisecs_time_to_datetime_object(self) -> None:
         msecs = 1690761600000
         dt = utils.convert_millisecs_time_to_datetime_object(msecs)

--- a/core/utils_test.py
+++ b/core/utils_test.py
@@ -770,12 +770,22 @@ class UtilsTests(test_utils.GenericTestBase):
     def test_get_time_in_millisecs(self) -> None:
         dt = datetime.datetime(2020, 6, 15)
         msecs = utils.get_time_in_millisecs(dt)
-        self.assertEqual(dt, datetime.datetime.fromtimestamp(msecs / 1000.0))
+        self.assertEqual(
+            dt,
+            datetime.datetime.fromtimestamp(
+                msecs / 1000.0, datetime.timezone.utc
+            ).replace(tzinfo=None),
+        )
 
     def test_get_time_in_millisecs_with_complicated_time(self) -> None:
         dt = datetime.datetime(2020, 6, 15, 5, 18, 23, microsecond=123456)
         msecs = utils.get_time_in_millisecs(dt)
-        self.assertEqual(dt, datetime.datetime.fromtimestamp(msecs / 1000.0))
+        self.assertEqual(
+            dt,
+            datetime.datetime.fromtimestamp(
+                msecs / 1000.0, datetime.timezone.utc
+            ).replace(tzinfo=None),
+        )
 
     def test_convert_millisecs_time_to_datetime_object(self) -> None:
         msecs = 1690761600000

